cloud/services/availabilitysets/availabilitysets.go

```diff
@@ -0,0 +1,132 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package availabilitysets
+
+import (
+	"context"
+	"strconv"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2020-06-30/compute"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+
+	"k8s.io/utils/pointer"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
+	azure "sigs.k8s.io/cluster-api-provider-azure/cloud"
+	"sigs.k8s.io/cluster-api-provider-azure/cloud/converters"
+	"sigs.k8s.io/cluster-api-provider-azure/cloud/services/resourceskus"
+	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
+)
+
+// AvailabilitySetScope defines the scope interface for a availability sets service.
+type AvailabilitySetScope interface {
+	logr.Logger
+	azure.ClusterDescriber
+	AvailabilitySetSpecs() []azure.AvailabilitySetSpec
+}
+
+// Service provides operations on azure resources
+type Service struct {
+	Scope AvailabilitySetScope
+	Client
+	resourceSKUCache *resourceskus.Cache
+}
+
+// New creates a new availability sets service.
+func New(scope AvailabilitySetScope, skuCache *resourceskus.Cache) *Service {
+	return &Service{
+		Scope:            scope,
+		Client:           NewClient(scope),
+		resourceSKUCache: skuCache,
+	}
+}
+
+// Reconcile creates or updates availability sets
+func (s *Service) Reconcile(ctx context.Context) error {
+	ctx, span := tele.Tracer().Start(ctx, "availabilitysets.Service.Reconcile")
+	defer span.End()
+
+	asSku, err := s.resourceSKUCache.Get(ctx, string(compute.Aligned), resourceskus.AvailabilitySets)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get availability sets sku")
+	}
+
+	maxFaultDomainCountStr, ok := asSku.GetCapability(resourceskus.MaximumPlatformFaultDomainCount)
+	if !ok {
+		return errors.Errorf("cannot find capability %s sku %s", resourceskus.MaximumPlatformFaultDomainCount, *asSku.Name)
+	}
+
+	maxFaultDomainCount, err := strconv.ParseUint(maxFaultDomainCountStr, 10, 32)
+	if err != nil {
+		return errors.Wrapf(err, "failed to determine max fault domain count")
+	}
+
+	for _, asSpec := range s.Scope.AvailabilitySetSpecs() {
+		s.Scope.V(2).Info("creating availability set", "availability set", asSpec.Name)
+
+		faultDomainCount := asSpec.DesiredFaultDomainCount
+		if asSpec.DesiredFaultDomainCount > int32(maxFaultDomainCount) {
+			s.Scope.V(2).Info("desired fault domain count is greater than max fault domain count, will use the max fault domain count",
+				"desired fault domain count", asSpec.DesiredFaultDomainCount, "max fault domain count", maxFaultDomainCount)
+			faultDomainCount = int32(maxFaultDomainCount)
```

Comment:
What's the point of `DesiredFaultDomainCount`? Why not always use `maxFaultDomainCount` ? Do we ever want to use a lower number than max?

Reply:
Not now, just wanted to keep `AvailabilitySetSpec` open for changes in future. I'm okay with removing it and hardcoding it to max count if it make it simple.